opslevel/resource_opslevel_infra.go

```diff
@@ -1,219 +1,303 @@
+	}
+
+	infrastructure, err := r.client.GetInfrastructure(data.Id.ValueString())
+	if err != nil {
+		resp.Diagnostics.AddError("opslevel client error", fmt.Sprintf("Unable to read infrastructure, got error: %s", err))
+		return
+	}
+	readInfrastructureResourceModel, diags := NewInfrastructureResourceModel(ctx, *infrastructure)
+	resp.Diagnostics.Append(diags...)
+	readInfrastructureResourceModel.Aliases = data.Aliases
+
+	// Save updated data into Terraform state
+	resp.Diagnostics.Append(resp.State.Set(ctx, &readInfrastructureResourceModel)...)
+}
+
+func (r *InfrastructureResource) Update(ctx context.Context, req resource.UpdateRequest, resp *resource.UpdateResponse) {
+	var data InfrastructureResourceModel
+
+	// Read Terraform plan data into the model
+	resp.Diagnostics.Append(req.Plan.Get(ctx, &data)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	infraInput, err := newInfraInput(data)
+	if err != nil {
+		resp.Diagnostics.AddError("opslevel client error", fmt.Sprintf("Unable to create opslevel InfraInput, got error: %s", err))
+		return
+	}
+	updatedInfrastructure, err := r.client.UpdateInfrastructure(data.Id.ValueString(), infraInput)
+	if err != nil {
+		resp.Diagnostics.AddError("opslevel client error", fmt.Sprintf("Unable to update infrastructure, got error: %s", err))
+		return
+	}
+	updatedInfrastructureResourceModel, diags := NewInfrastructureResourceModel(ctx, *updatedInfrastructure)
+	resp.Diagnostics.Append(diags...)
+	updatedInfrastructureResourceModel.Aliases = data.Aliases
+	updatedInfrastructureResourceModel.LastUpdated = timeLastUpdated()
+
+	tflog.Trace(ctx, "updated a infrastructure resource")
+	resp.Diagnostics.Append(resp.State.Set(ctx, &updatedInfrastructureResourceModel)...)
+}
+
+func (r *InfrastructureResource) Delete(ctx context.Context, req resource.DeleteRequest, resp *resource.DeleteResponse) {
+	var data InfrastructureResourceModel
+
+	// Read Terraform prior state data into the model
+	resp.Diagnostics.Append(req.State.Get(ctx, &data)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	err := r.client.DeleteInfrastructure(data.Id.ValueString())
+	if err != nil {
+		resp.Diagnostics.AddError("Client Error", fmt.Sprintf("Unable to delete infrastructure, got error: %s", err))
+		return
+	}
+	tflog.Trace(ctx, "deleted a infrastructure resource")
+}
+
+func (r *InfrastructureResource) ImportState(ctx context.Context, req resource.ImportStateRequest, resp *resource.ImportStateResponse) {
+	resource.ImportStatePassthroughID(ctx, path.Root("id"), req, resp)
+}
+
+func newInfraInput(infraModel InfrastructureResourceModel) (opslevel.InfraInput, error) {
+	infraInput := opslevel.InfraInput{Schema: infraModel.Schema.ValueString()}
+
+	if infraModel.Owner.IsNull() {
+		infraInput.Owner = opslevel.NewID("")
+	} else {
+		infraInput.Owner = opslevel.NewID(infraModel.Owner.ValueString())
+	}
+
+	if infraModel.Data.IsNull() {
+		// Unsets this previously set field
+		newJSON, err := opslevel.NewJSON("{}")
```

Comment:
Can you elaborate on what `"{}"` will do? 

```
type InfraInput struct {
	Schema   string              `json:"schema" yaml:"schema" default:"Database"`
	Owner    *ID                 `json:"owner,omitempty" yaml:"owner,omitempty" default:"XXX_owner_id_XXX"`
	Provider *InfraProviderInput `json:"provider" yaml:"provider"`
	Data     *JSON               `json:"data" yaml:"data" default:"{\"name\":\"my-big-query\",\"engine\":\"BigQuery\",\"endpoint\":\"https://google.com\",\"replica\":false}"`
}
```

Could we just exclude the `Data` field if it was not set?

Reply:
In this case, we want to send the API `"{}"` (on json marshal) and the API will deal with it correctly.